hphp/runtime/vm/jit/smashable-instr-arm.cpp

```diff
@@ -208,60 +231,85 @@ uint32_t smashableCmpqImm(TCA inst) {
 
-  // This doesn't verify that each of the two or three instructions that make
-  // up this sequence matches; just the first one and the indirect jump.
   Instruction* ldr = Instruction::Cast(inst);
-  if (ldr->Bits(31, 24) != 0x58) return nullptr;
-
-  Instruction* br = Instruction::Cast(inst + 4);
-  if (br->Bits(31, 10) != 0x3587C0 || br->Bits(4, 0) != 0) return nullptr;
-
-  uintptr_t dest = reinterpret_cast<uintptr_t>(inst + 8);
-  assertx((dest & 3) == 0);
-
-  return *reinterpret_cast<TCA*>(dest);
+  Instruction* br = ldr->NextInstruction();
+  Instruction* target = br->NextInstruction();
+  const auto rd = ldr->Rd();
+
+  if (ldr->IsLoadLiteral() &&
+      ldr->Mask(LoadLiteralMask) == LDR_x_lit &&
+      ldr->ImmPCOffsetTarget() == target &&
+      br->Mask(UnconditionalBranchToRegisterMask) == BR &&
+      br->Rn() == rd) {
+    assertx((reinterpret_cast<uintptr_t>(target) & 3) == 0);
+    return *reinterpret_cast<TCA*>(target);
+  }
+  return nullptr;
 }
 
 TCA smashableJccTarget(TCA inst) {
   using namespace vixl;
 
   Instruction* b = Instruction::Cast(inst);
-  if (b->Bits(31, 24) != 0x54 || b->Bit(4) != 0) return nullptr;
-
-  Instruction* br = Instruction::Cast(inst + 8);
-  if (br->Bits(31, 10) != 0x3587C0 || br->Bits(4, 0) != 0) return nullptr;
-
-  uintptr_t dest = reinterpret_cast<uintptr_t>(inst + 12);
-  assertx((dest & 3) == 0);
-
-  return *reinterpret_cast<TCA*>(dest);
+  Instruction* ldr = b->NextInstruction();;
+  Instruction* br = ldr->NextInstruction();
+  Instruction* target = br->NextInstruction();
+  Instruction* after = target->NextInstruction()->NextInstruction();
+  const auto rd = ldr->Rd();
+
+  if (b->IsCondBranchImm() &&
+      b->ImmPCOffsetTarget() == after &&
+      ldr->IsLoadLiteral() &&
+      ldr->Mask(LoadLiteralMask) == LDR_x_lit &&
+      ldr->ImmPCOffsetTarget() == target &&
+      br->Mask(UnconditionalBranchToRegisterMask) == BR &&
+      br->Rn() == rd) {
+    assertx((reinterpret_cast<uintptr_t>(target) & 3) == 0);
+    return *reinterpret_cast<TCA*>(target);
+  }
+  return nullptr;
 }
 
 ConditionCode smashableJccCond(TCA inst) {
   using namespace vixl;
-  Instruction* b = Instruction::Cast(inst);
-  if (b->Bits(31, 24) != 0x54 || b->Bit(4) != 0) return CC_NP;
 
-  Instruction* br = Instruction::Cast(inst + 8);
-  if (br->Bits(31, 10) != 0x3587C0 || br->Bits(4, 0) != 0) return CC_NP;
+  Instruction* b = Instruction::Cast(inst);
+  Instruction* ldr = b->NextInstruction();;
+  Instruction* br = ldr->NextInstruction();
+  Instruction* target = br->NextInstruction();
+  Instruction* after = target->NextInstruction()->NextInstruction();
+  const auto rd = ldr->Rd();
+
+  assertx(b->IsCondBranchImm() &&
+          b->ImmPCOffsetTarget() == after &&
+          ldr->IsLoadLiteral() &&
+          ldr->Mask(LoadLiteralMask) == LDR_x_lit &&
+          ldr->ImmPCOffsetTarget() == target &&
+          br->Mask(UnconditionalBranchToRegisterMask) == BR &&
+          br->Rn() == rd);
```

Comment:
Can all of these changes be pulled into a separate PR?

The more you can split this up, the better.